File: solver/src/sudoku.cpp
#include <vector>
#include <string>
#include <iostream>
#include <algorithm>
#include <unordered_map>
#include <set>
#include <memory>
#include "solver/sudoku.h"
#include "solver/variables.h"

using std::cout;
using std::endl;
using std::string;
using std::vector;

namespace Sudoku
{

std::vector<int> Solver::Solve(const std::vector<int>& grid) {
    auto game = std::make_shared<Game>(data_.cells);
    // init game with cells
    bool success = SetupGameWithGrid(grid, game);
    if (!success) {
        std::cout << "could not setup game! please check your grid." << std::endl;
        return grid; // Return original grid if setup fails
    }
     
    // search all cells, returns solved solution
    std::shared_ptr<Game> finished = Search(game);

    // Convert the solution back to vector<int> format
    std::vector<int> solution(81);
    int i = 0;
    for (auto& sq : data_.squares) {
        solution[i++] = finished->cells[sq].candidates[0]; // Get the single remaining candidate
    }
    
    return solution;
}

bool Solver::SetupGameWithGrid(const std::vector<int>& grid, std::shared_ptr<Game>& game) {
    // To start, every square can be any digit; then assign values from the grid.
    bool success = true;
    for (int i = 0; i < data_.squares.size(); i++) {
        if (grid[i] != 0) {
            if (!Assign(game, data_.squares[i], grid[i])) {
                success = false;
            }
        }
    }
    return success;
}

bool Solver::Assign(std::shared_ptr<Sudoku::Game>& game, std::string key, int digit) {
    Cell& cell = game->cells[key];
    PrintEvent("assigning", 1, digit, cell);
    
    // get all values besides possible candidates that work with that cell
    GetAllValuesExcept(cell.candidates, digit);

    // Eliminate all other values that aren't the assign value from candidate
    for (auto& candidate : cell.candidates) {
        // we're asking if this 
        if (!Eliminate(game, key, candidate)) {
            return false;
        }
    } 
    return true;
}

void Solver::GetAllValuesExcept(std::vector<int>& others, int to_remove) {
    others.erase(
        std::remove_if(others.begin(), others.end(),
        [&](int& s) { return s == to_remove; }),    
    others.end());
}

void Solver::PrintEvent(const std::string& message, const int level, const int digit, Cell& cell) {
    std::string tabLevel(level, '\t');
    std::cout << tabLevel << message << ": ";
    std::cout << "cell: " << cell.key << " ";
    std::cout << "digit: " << digit << " ";
    std::cout << "values[s]: ";
    for (auto& c : cell.candidates) {
        std::cout << c;
    }
    std::cout << std::endl;
}

bool Solver::Eliminate(std::shared_ptr<Game>& game, std::string key, int digit) {
    // if digit doesn't exist in cell value candidates return 
    Cell& cell = game->cells[key];
    if (!cell.HasCandidate(digit)) {
        return true;
    }
    cell.RemoveCandidate(digit);
    PrintEvent("eliminated", 1, digit, cell);

    // if cell candidate is greater than 1, remove from candidates 
    if (cell.candidates.empty()) {
        PrintEvent("contradiction:eliminate", 0, digit, cell);
        return false;
    } else if (cell.candidates.size() == 1) {
        PrintEvent("eliminate.peers", 2, digit, cell);
        for (auto peer : cell.peers) {
            if (digit != 0 && cell.candidates.size() > 0) {
                // std::cout << "eliminate d:" << digit << " from cell: " << data_.cells[peer].key << " which is a peer of " << cell.key << std::endl;
                if (!Eliminate(game, peer, cell.candidates[0])) {
                    std::cout << "eliminate 2 returned false!" << std::endl;
                    return false;
                }
            }
        }
    }
    // by using process of elimination, for squares that have one possible value
    // see if we can see if we can assign a value there
    for (auto& unit : cell.units) {
        std::vector<std::string> dplaces;
        for (auto& v : unit) {
            if (game->cells[v].HasCandidate(digit)) {
                dplaces.push_back(v);
            }
        }
        if (dplaces.empty()) {
            return false;
        }
        if (dplaces.size() == 1) {
            if (!Assign(game, dplaces[0], digit)) {
                return false;
            }
        }
    }
    // if it did not fail and return, it must have worked
    return true;
}

std::shared_ptr<Game> Solver::Search(std::shared_ptr<Game>& game) {
    if (isSolved(game->cells)) {
        game->state = GameState::complete;
        return game;
    }

    if (game->state == GameState::failed) {
        return game;
    }

    // get unordered_map with key of sq and value of cell candidate size
    std::unordered_map<std::string, int> pos;
    for (auto& s : data_.squares) {
        Cell cell = game->cells[s];
        if (cell.candidates.size() > 1) {
           pos[cell.key] = cell.candidates.size();
        }
    }
    
    // get keys for possible undetermined s values
    std::vector<std::string> res; 
    res.reserve(pos.size());
    for (auto& p : pos) {
        res.push_back(p.first);
    }
   
    // then sort res vector
    std::sort(res.begin(), res.end(), 
        [&](std::string a, std::string b) {
            if (game->cells[a].candidates.size() == game->cells[b].candidates.size()) {
                return a < b;
            } else {
                return game->cells[a].candidates.size() < game->cells[b].candidates.size();
            }
        });

    // then get the key of square w/ smallest candidate.size()
    std::string min_cand_key = res[0];

    // Try each value for the selected variable
    for (auto& p : game->cells[min_cand_key].candidates) {
        auto game_to_test = std::make_shared<Game>(game);
        if (Assign(game_to_test, min_cand_key, p)) {
            auto result = Search(game_to_test);
            if (result->state != GameState::failed) {
                return result;  // Return immediately if we found a solution
            }
        }
    }

    // If we get here, no solution was found
    game->state = GameState::failed;
    return game;
}

bool Solver::isSolved(std::unordered_map<std::string , Cell>& cells) {
    bool solved = true;
    for (auto& c : cells) {
        if (c.second.candidates.size() > 1) {
            solved = false;
            break;
        }
    }
    return solved;
}

bool Solver::ValidateGrid(const std::vector<int>& grid, bool debug) {
    // Check grid size
    if (grid.size() != 81) {
        if (debug) std::cout << "Invalid grid size: " << grid.size() << std::endl;
        return false;
    }

    // Check for valid digits (0-9 only)
    for (int digit : grid) {
        if (digit < 0 || digit > 9) {
            if (debug) std::cout << "Invalid digit found: " << digit << std::endl;
            return false;
        }
    }

    // Check rows, columns, and boxes for duplicate numbers
    for (int i = 0; i < 9; i++) {
        std::vector<bool> row(10, false);
        std::vector<bool> col(10, false);
        std::vector<bool> box(10, false);

        for (int j = 0; j < 9; j++) {
            // Check row
            int rowVal = grid[i * 9 + j];
            if (rowVal != 0) {
                if (row[rowVal]) {
                    if (debug) std::cout << "Duplicate " << rowVal << " in row " << i << std::endl;
                    return false;
                }
                row[rowVal] = true;
            }

            // Check column
            int colVal = grid[j * 9 + i];
            if (colVal != 0) {
                if (col[colVal]) {
                    if (debug) std::cout << "Duplicate " << colVal << " in column " << i << std::endl;
                    return false;
                }
                col[colVal] = true;
            }

            // Check box
            int boxRow = 3 * (i / 3) + j / 3;
            int boxCol = 3 * (i % 3) + j % 3;
            int boxVal = grid[boxRow * 9 + boxCol];
            if (boxVal != 0) {
                if (box[boxVal]) {
                    if (debug) std::cout << "Duplicate " << boxVal << " in box " << i << std::endl;
                    return false;
                }
                box[boxVal] = true;
            }
        }
    }

    return true;
}

Solver::~Solver() {}

} // namespace Sudoku

File: sudoku.py
from typing import List, Dict, Set
from dataclasses import dataclass
from enum import Enum
import copy
from sudoku_net.src.utils.logging_config import setup_logger

# Setup logging configuration
logger = setup_logger(__name__)

class GameState(Enum):
    in_progress = "in_progress"
    complete = "complete"
    failed = "failed"

@dataclass
class Cell:
    key: str
    candidates: List[int]
    peers: Set[str]
    units: List[List[str]]

    def has_candidate(self, digit: int) -> bool:
        return digit in self.candidates

    def remove_candidate(self, digit: int) -> None:
        if digit in self.candidates:
            self.candidates.remove(digit)

class Game:
    def __init__(self, cells: Dict[str, Cell]):
        self.cells = copy.deepcopy(cells)
        self.state = GameState.in_progress

class Variables:
    def __init__(self):
        self.digits = list(range(1, 10))
        self.rows = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I']
        self.cols = ['1', '2', '3', '4', '5', '6', '7', '8', '9']
        self.squares = self.cross(self.rows, self.cols)
        self.cells: Dict[str, Cell] = {}
        
        unitlist = []
        self.setup_unitlist(unitlist)
        self.setup_cells(unitlist)
        self.setup_peers()

    def cross(self, A: List[str], B: List[str]) -> List[str]:
        return [a + b for a in A for b in B]

    def setup_unitlist(self, unitlist: List[List[str]]) -> None:
        # Column-wise units
        col_cross = [[r + c for r in self.rows] for c in self.cols]
        
        # Row-wise units
        row_cross = [[r + c for c in self.cols] for r in self.rows]
        
        # Box units
        row_units = [self.rows[i:i + 3] for i in range(0, 9, 3)]
        col_units = [self.cols[i:i + 3] for i in range(0, 9, 3)]
        
        unit_cross = []
        for ru in row_units:
            for cu in col_units:
                unit_cross.append([r + c for r in ru for c in cu])
        
        unitlist.extend(col_cross)
        unitlist.extend(row_cross)
        unitlist.extend(unit_cross)

    def setup_cells(self, unitlist: List[List[str]]) -> None:
        for k in self.squares:
            cell = Cell(key=k, candidates=list(self.digits), peers=set(), units=[])
            for unit in unitlist:
                if k in unit:
                    cell.units.append(unit)
            self.cells[k] = cell

    def setup_peers(self) -> None:
        for k in self.squares:
            peers = set()
            for unit in self.cells[k].units:
                for v in unit:
                    if v != k:
                        peers.add(v)
            self.cells[k].peers = peers

class Solver:
    def __init__(self):
        self.data = Variables()

    def solve(self, grid: List[int]) -> None:
        game = Game(self.data.cells)
        success = self.setup_game_with_grid(grid, game)
        
        if not success:
            logger.error("Could not setup game! Please check your grid.")
            return
        
        finished = self.search(game)
        return finished

    def setup_game_with_grid(self, grid: List[int], game: Game) -> bool:
        success = True
        for i, digit in enumerate(grid):
            if digit != 0:
                square = self.data.squares[i]
                if not self.assign(game, square, digit):
                    success = False
        return success

    def assign(self, game: Game, key: str, digit: int) -> bool:
        cell = game.cells[key]
        others = [d for d in cell.candidates if d != digit]
        
        for other in others:
            if not self.eliminate(game, key, other):
                return False
        return True

    def eliminate(self, game: Game, key: str, digit: int) -> bool:
        cell = game.cells[key]
        if not cell.has_candidate(digit):
            return True
            
        cell.remove_candidate(digit)

        if len(cell.candidates) == 0:
            return False
        elif len(cell.candidates) == 1:
            value = cell.candidates[0]
            for peer in cell.peers:
                if not self.eliminate(game, peer, value):
                    return False

        for unit in cell.units:
            dplaces = [s for s in unit if game.cells[s].has_candidate(digit)]
            if len(dplaces) == 0:
                return False
            elif len(dplaces) == 1:
                if not self.assign(game, dplaces[0], digit):
                    return False
        return True

    def search(self, game: Game) -> Game:
        if self.is_solved(game.cells):
            game.state = GameState.complete
            logger.info("Solved!")
            return game

        if game.state == GameState.failed:
            return game

        # Find the square with the fewest candidates
        n, s = min((len(game.cells[s].candidates), s) 
                   for s in self.data.squares 
                   if len(game.cells[s].candidates) > 1)

        for d in game.cells[s].candidates:
            game_copy = Game(game.cells)
            if self.assign(game_copy, s, d):
                result = self.search(game_copy)
                if result.state != GameState.failed:
                    return result
        
        game.state = GameState.failed
        return game

    @staticmethod
    def is_solved(cells: Dict[str, Cell]) -> bool:
        return all(len(cell.candidates) == 1 for cell in cells.values())